wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/CheEnvironmentValidator.java

```diff
@@ -33,92 +45,269 @@
+                                      .map(Map.Entry::getKey)
+                                      .collect(toList());
+
+        checkArgument(devMachines.size() == 1,
+                      "Environment '%s' should contain exactly 1 machine with ws-agent, but contains '%d'. " +
+                      "All machines with this agent: %s",
+                      envName, devMachines.size(), Joiner.on(", ").join(devMachines));
+
+        // needed to validate different kinds of dependencies in services to other services
+        Set<String> servicesNames = composeEnvironment.getServices().keySet();
+
+        composeEnvironment.getServices()
+                          .forEach((serviceName, service) -> validateMachine(serviceName,
+                                                                             env.getMachines().get(serviceName),
+                                                                             service,
+                                                                             envName,
+                                                                             servicesNames));
+
+        // check that order can be resolved
+        try {
+            startStrategy.order(composeEnvironment);
+        } catch (IllegalArgumentException e) {
+            throw new IllegalArgumentException(
+                    format("Compose services order of environment '%s' is not resolvable. Error: %s",
+                           envName, e.getLocalizedMessage()));
         }
     }
 
-    private void validateMachine(MachineConfig machineCfg, String envName) throws IllegalArgumentException {
+    protected void validateMachine(String machineName,
+                                   @Nullable ExtendedMachine extendedMachine,
+                                   ComposeService service,
+                                   String envName,
+                                   Set<String> servicesNames) throws IllegalArgumentException {
+        checkArgument(MACHINE_NAME_PATTERN.matcher(machineName).matches(),
+                      "Name of machine '%s' in environment '%s' is invalid",
+                      machineName, envName);
+
+        checkArgument(!isNullOrEmpty(service.getImage()) ||
+                      (service.getBuild() != null && !isNullOrEmpty(service.getBuild().getContext())),
+                      "Filed 'image' or 'build.context' is missing in compose service '%s' in environment '%s'",
+                      machineName, envName);
+
+        checkArgument(service.getMemLimit() != null && service.getMemLimit() > 0,
+                      "Field 'mem_limit' is missing or contain illegal value in service '%s' of environment '%s'." +
+                      " This field should specify container RAM size in bytes.",
+                      machineName, envName);
+
+        service.getExpose()
+               .stream()
+               .forEach(expose -> checkArgument(EXPOSE_PATTERN.matcher(expose).matches(),
```

Comment:
Why not to use native foreach?, it looks better for me + i think it will work faster and will use less memory as you don't need to create stream for each collection you'r checking